File: src/main/java/victor/training/performance/spring/kt/Calculator.java
package victor.training.performance.spring.kt;

import java.util.HashSet;
import java.util.Set;

public class Calculator {
    interface Listener {
        void onResult(int result);
    }
    Set<Listener> listeners = new HashSet<>();
    void addListener(Listener listener) {
        listeners.add(listener);
    }
    void removeListener(Listener listener) {
        listeners.remove(listener);
    }

    public int sum(int a, int b) {

        int s = a + b;
        for (Listener listener : listeners) {
            listener.onResult(s);
        }
        return s;
    }
}

File: src/main/java/victor/training/performance/spring/kt/Calculator.kt
package victor.training.performance.spring.kt

typealias Listener = (Int) -> Unit
class Calculator {
    var listeners: MutableSet<Listener> = HashSet()

    fun addListener(listener: Listener) {
        listeners.add(listener)
    }

    fun removeListener(listener: Listener) {
        listeners.remove(listener)
    }

    fun sum(a: Int, b: Int): Int {
        val s = a + b
        for (listener in listeners) {
            listener(s)
        }
        return s
    }
}